### PR TITLE
Add delegates route

### LINKED ISSUE #101: Add GET Delegates
- Add `/v1/chains/<chain_id>/delegates?<safe>&<delegate>&<delegator>&<label>` route
- Tests should be added
- OpenApi specification should be added

### PR DESCRIPTION
**Pull request by moisses89**
Closes #101 
- Add the endpoint GET delegates
- Add some tests

### CODE REVIEW

File: src/routes/delegates/delegates.controller.ts [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    required: false,
+    type: String,
+  })
+  @ApiImplicitQuery({
+    name: 'cursor',
+    required: false,
+    type: String,
+  })
+  @Get('chains/:chainId/delegates')
+  async getDelegates(
```
Comment (moisses89) [inline src/routes/delegates/delegates.controller.ts:46]: We should validate here that at least one query parameter between (safe, delegate, delegator or label) is not `null` because if not, the `tx-service` returns `400 bad request `

File: src/domain.module.ts [hunk 2]
```diff
@@ -66,6 +66,7 @@ import { DelegateRepository } from './domain/delegate/delegate.repository';
     IExchangeRepository,
     ISafeRepository,
     IContractsRepository,
+    IDelegateRepository,
```
Comment (fmrsabino) [inline src/domain.module.ts:69]: Please use lexicographical order

File: src/routes/delegates/entities/delegate.entity.ts [hunk 3]
```diff
@@ -0,0 +1,12 @@
+import { ApiProperty } from '@nestjs/swagger';
+
+export class Delegate {
+  @ApiProperty()
```
Comment (fmrsabino) [inline src/routes/delegates/entities/delegate.entity.ts:4]: You should ApiPropertyOptional for optional fields

File: src/routes/delegates/entities/delegate.entity.ts [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+  safe?: string;
+  @ApiProperty()
+  delegate: string;
+  @ApiProperty()
+  delegator: string;
+  @ApiProperty()
+  label: string;
```
Comment (fmrsabino) [inline src/routes/delegates/entities/delegate.entity.ts:11]: Shouldn't all of these fields be optional?

File: src/domain/delegate/entities/__tests__/delegate.factory.ts [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+  delegator?: string,
+  label?: boolean,
+): Delegate {
+  return <Delegate>{
+    safe: safe ?? faker.random.word(),
+    delegate: delegate ?? faker.random.word(),
+    delegator: delegator ?? faker.random.word(),
+    label: label ?? faker.random.word(),
+  };
+}
```
Comment (fmrsabino) [inline src/domain/delegate/entities/__tests__/delegate.factory.ts:15]: You should use faker.finance.ethereumAddress() for these fields

File: src/routes/delegates/delegates.controller.spec.ts [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+
+    app = moduleFixture.createNestApplication();
+    app.useGlobalFilters(new DataSourceErrorFilter());
+
+    await app.init();
+  });
+
+  describe('GET delegates for a Safe', () => {
+    it('Success', async () => {
+      const safe = '0x5afe3855358e112b5647b952709e6165e1c1eeee';
```
Comment (fmrsabino) [inline src/routes/delegates/delegates.controller.spec.ts:65]: nitpick: you can generate a random ethereum address with faker

File: src/routes/delegates/delegates.controller.spec.ts [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+        params: {
+          safe: safe,
+          delegate: undefined,
+          delegator: undefined,
+          label: undefined,
+          limit: undefined,
+          offset: undefined,
+        },
+      });
+    });
```
Comment (fmrsabino) [inline src/routes/delegates/delegates.controller.spec.ts:97]: From the discussion that we had yesterday I'd say that you can skip these checks at this level.

We have two different takes for this one:
- They have some value in validating that we actually call external services with these parameters (testing the boundaries)
- They add complexity to the test which makes it a little bit harder to read.

I have no strong opinion on this one but I think given the feedback that I got I'm leaning towards what the team finds more readable.

File: src/routes/delegates/delegates.controller.spec.ts [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+
+    it('Should failure with bad request', async () => {
```
Comment (fmrsabino) [inline src/routes/delegates/delegates.controller.spec.ts:134]: nitpick: "Should fail"

File: src/routes/delegates/delegates.service.spec.ts [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+    delegateRepositoryMock,
+  );
+
+  beforeEach(async () => {
+    jest.clearAllMocks();
+  });
+
+  it('should retrieve the delegates from a safe', async () => {
+    const chainId = '1';
+    const safe = '0x5afe3855358e112b5647b952709e6165e1c1eeee';
```
Comment (fmrsabino) [inline src/routes/delegates/delegates.service.spec.ts:25]: nitpick: you can generate a random ethereum address with faker

File: src/routes/delegates/delegates.service.ts [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+      delegates.previous,
+    );
+
+    return <Page<Delegate>>{
+      count: delegates.count,
+      next: nextURL?.toString(),
+      previous: previousURL?.toString(),
+      results: delegates.results,
+    };
+  }
```
Comment (fmrsabino) [inline src/routes/delegates/delegates.service.ts:48]: We should consider returning the OpenApi entity at this level. Currently it works because the interface matches.

For this we would need to add a constructor to the Page entity on the route level (which should be done a in a separate PR).

**Review by fmrsabino — CHANGES_REQUESTED** [review]

File: src/routes/delegates/delegates.controller.spec.ts (hunk 7 above)
Comment (moisses89) [reply src/routes/delegates/delegates.controller.spec.ts:97]: Ok, removing the more complex checks.

File: src/routes/delegates/delegates.controller.spec.ts (hunk 8 above)
Comment (moisses89) [reply src/routes/delegates/delegates.controller.spec.ts:134]: @fmrsabino should return a bad request, and returns 503 something is wrong

File: src/routes/delegates/delegates.controller.ts (hunk 1 above)
Comment (hectorgomezv) [reply src/routes/delegates/delegates.controller.ts:46]: If we agree on using TS guards as validation tool for request, you can find [here](https://github.com/5afe/safe-client-gateway-nest/pull/132/commits/206ab186563e3f3945facd881c6da292aca9231e#diff-12bd9e35019ab1c8ce38b567178f33911bc1253984905fb8b274f6779a804d8a) an example (pending to be reviewed). 

In my opinion would be nice to map the values in a DTO and validate it, but you could also implement the validation at `DelegatesService` level manually, no strong opinion here. And probably we could do it in a separate issue/PR as well if you prefer.

**Review by hectorgomezv — DISMISSED** [review]
LGTM, just left a minor comment on validation. Feel free to implement it here or we can create another issue/PR.

File: src/routes/delegates/delegates.service.ts (hunk 10 above)
Comment (hectorgomezv) [reply src/routes/delegates/delegates.service.ts:48]: My two cents: before returning OpenApi entities from services, I think we should think if _services_ do really belong to the _routes_ layer. I think we can all agree OpenApi definitions belong to _routes_ (they're REST-specific), but do the _services_ belong to _routes_? As I said in a previous PR, if we integrate another presentation technology (websockets, cli, gRPC, whatever) I'd expect to have _services_ available in _domain_ 🤔

Comment (fmrsabino) [reply src/routes/delegates/delegates.service.ts:48]: Services do the "business" logic associated with the route ie.: presentation layer. The term itself shouldn't be a big concern – it was taken from NestJS.

If we move the Service to the domain:
1.  What is the role of the service in the domain?
2. How do you encapsulate business logic for the route?

File: src/routes/delegates/delegates.controller.spec.ts (hunk 8 above)
Comment (fmrsabino) [reply src/routes/delegates/delegates.controller.spec.ts]: > @fmrsabino should return a bad request, and returns 503 something is wrong

Your comment is unrelated with my comment right? Maybe I misunderstood. But are you saying that it should return 400 instead of 503 in this test?

File: src/routes/delegates/delegates.service.ts (hunk 10 above)
Comment (hectorgomezv) [reply src/routes/delegates/delegates.service.ts:48]: After an offline talk, I think both of us agree this possible refactor is not needed right now, and we should focus in having feature parity with the current CGW. I'll clarify my concerns/proposal in the future 🙂 thanks.

File: src/routes/delegates/entities/delegate.entity.ts (hunk 4 above)
Comment (fmrsabino) [reply src/routes/delegates/entities/delegate.entity.ts:11]: Sorry yes you are right. I was mixing these parameters with the query ones.

**Comment by moisses89** [comment]
@fmrsabino  @hectorgomezv  Added validation of query parameters https://github.com/5afe/safe-client-gateway-nest/pull/133/commits/ca373efa3eafadba01813394915ab90945b5d80e

**Review by hectorgomezv — APPROVED** [review]

File: src/routes/delegates/entities/delegate-params.entity.ts [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+    this.delegator = delegator;
+    this.label = label;
+  }
+}
+
+export function isDelegateParamsDto(
+  dto: DelegateParamsDto,
+): dto is DelegateParamsDto {
+  return !!(dto.safe || dto.delegate || dto.delegator || dto.label);
+}
```
Comment (fmrsabino) [inline src/routes/delegates/entities/delegate-params.entity.ts:30]: I think this should not be used as a type predicate. If you look into `DelegateParamsDto` almost any TS object can matched with `DelegateParamsDto` because all the fields are nullable. So the question is not if it is a `DelegateParamsDto` but if at least one of the fields is not null.

So the following is fine:

``` typescript
if (
      !(
        delegateParams.safe ||
        delegateParams.delegate ||
        delegateParams.delegator ||
        delegateParams.label
      )
    ) {
      throw new HttpException(
        'At least one query param must be provided',
        HttpStatus.BAD_REQUEST,
      );
    }
```

Also sidenote: you have double negation `!!` in the original condition which is not needed.

**Review by fmrsabino — CHANGES_REQUESTED** [review]

File: src/routes/delegates/entities/delegate-params.entity.ts (hunk 11 above)
Comment (moisses89) [reply src/routes/delegates/entities/delegate-params.entity.ts:30]: Ok, agree maybe the function could be called isValid... instead isType. But is ok. 
Regarding double negation I was aware, because should return `true` when was valid. 
```
> !(("") || ("") || (null) )
true
> !!(("Hello") || ("") || (null) )
true
```

Comment (moisses89) [reply src/routes/delegates/entities/delegate-params.entity.ts]: Done [bfc29e8c8c0a08242b33faf5fcedf88679d43664](https://github.com/5afe/safe-client-gateway-nest/pull/133/commits/bfc29e8c8c0a08242b33faf5fcedf88679d43664)

**Review by fmrsabino — APPROVED** [review]
